File: samples/java/ShoppingMall/Server/Configuration/src/main/java/systems/zlink/samples/shoppingmall/server/configuration/SampleTopology.java
package systems.zlink.samples.shoppingmall.server.configuration;

import org.springframework.boot.context.properties.ConfigurationProperties;

@ConfigurationProperties("sample")
public record SampleTopology(
        String instanceName,
        String logDirectory,
        String httpUrl,
        String channelEndpoint,
        String spotEndpoint,
        String spotRouterEndpoint,
        String redisEndpoint,
        String redisKeyPrefix) {

    public Api api() {
        return new Api(
                required(instanceName, "instanceName"),
                required(logDirectory, "logDirectory"),
                required(httpUrl, "httpUrl"));
    }

    public Workflow workflow() {
        String name = required(instanceName, "instanceName");
        return new Workflow(
                name,
                required(logDirectory, "logDirectory"),
                required(httpUrl, "httpUrl"),
                required(channelEndpoint, "channelEndpoint"),
                required(spotEndpoint, "spotEndpoint"),
                required(spotRouterEndpoint, "spotRouterEndpoint"));
    }

    public Location location() {
        return new Location(
                required(redisEndpoint, "redisEndpoint"),
                required(redisKeyPrefix, "redisKeyPrefix"));
    }

    public static String configPath(String[] args) {
        if (args.length != 2 || !"--config".equals(args[0]) || args[1].isBlank()) {
            throw new IllegalArgumentException("Usage: <role executable> --config <path>");
        }
        return args[1];
    }

    private static String required(String value, String name) {
        if (value == null || value.isBlank())
            throw new IllegalArgumentException("sample." + name + " is required");
        return value;
    }

    public record Api(String instanceName, String logDirectory, String httpUrl) {}

    public record Workflow(
            String instanceName,
            String logDirectory,
            String httpUrl,
            String channelEndpoint,
            String spotEndpoint,
            String spotRouterEndpoint) {}

    public record Location(String redisEndpoint, String redisKeyPrefix) {}
}

File: samples/kotlin/ShoppingMall/Server/Configuration/src/main/kotlin/systems/zlink/samples/kotlin/shoppingmall/server/configuration/SampleTopology.kt
package systems.zlink.samples.kotlin.shoppingmall.server.configuration

import java.time.Duration
import org.springframework.boot.context.properties.ConfigurationProperties
import systems.zlink.framework.locations.redis.ZLinkRedisLocationOptions
import systems.zlink.framework.locations.redis.ZLinkRedisLocationStore
import systems.zlink.samples.kotlin.shoppingmall.shared.contracts.OrderState
import systems.zlink.samples.kotlin.shoppingmall.shared.contracts.OrderStatuses

@ConfigurationProperties("sample")
data class SampleTopology(
    val instanceId: String? = null,
    val httpEndpoint: String? = null,
    val logDirectory: String? = null,
    val channelEndpoint: String? = null,
    val redisEndpoint: String? = null,
    val redisKeyPrefix: String? = null,
    val storeDirectory: String? = null,
) {
    fun role(): Role =
        Role(
            required(instanceId, "instanceId"),
            required(httpEndpoint, "httpEndpoint"),
            required(logDirectory, "logDirectory"),
            required(channelEndpoint, "channelEndpoint"),
            required(storeDirectory, "storeDirectory"),
        )

    fun location(): Location =
        Location(
            required(redisEndpoint, "redisEndpoint"),
            required(redisKeyPrefix, "redisKeyPrefix"),
        )

    fun requiredStoreDirectory(): String = required(storeDirectory, "storeDirectory")

    fun failedPlaceholder(orderId: String): OrderState =
        OrderState(
            orderId,
            OrderStatuses.Failed,
            null,
            null,
            null,
            "order does not exist",
            null,
            null,
            System.currentTimeMillis(),
        )

    data class Role(
        val instanceId: String,
        val httpEndpoint: String,
        val logDirectory: String,
        val channelEndpoint: String,
        val storeDirectory: String,
    )

    data class Location(val redisEndpoint: String, val redisKeyPrefix: String)

    companion object {
        fun configPath(args: Array<String>): String {
            require(args.size == 2 && args[0] == "--config" && args[1].isNotBlank()) {
                "Usage: <role executable> --config <path>"
            }
            return args[1]
        }

        private fun required(value: String?, name: String): String {
            require(!value.isNullOrBlank()) { "sample.$name is required" }
            return value
        }
    }
}

object SampleLocationStore {
    fun create(topology: SampleTopology): ZLinkRedisLocationStore {
        val location = topology.location()
        return ZLinkRedisLocationStore(
            ZLinkRedisLocationOptions()
                .setConnectionString(location.redisEndpoint)
                .setKeyPrefix("${location.redisKeyPrefix}locations:")
                .setCommandTimeout(Duration.ofMillis(500))
        )
    }
}
